File: CleanCodeJava/app/src/main/java/io/chuumong/cleancodejava/di/main/MainModule.java
package io.chuumong.cleancodejava.di.main;

import dagger.Module;
import dagger.Provides;
import io.chuumong.cleancodejava.data.api.CityService;
import io.chuumong.cleancodejava.data.repository.MainRepository;
import io.chuumong.cleancodejava.di.ActivityScope;
import io.chuumong.cleancodejava.presenter.MainPresenter;
import io.chuumong.cleancodejava.presenter.MainPresenterImpl;
import io.chuumong.cleancodejava.presenter.MainView;

/**
 * Created by JongHunLee on 2017-02-22.
 */

@Module
public class MainModule {

    private final MainView view;

    public MainModule(MainView view) {
        this.view = view;
    }

    @Provides
    @ActivityScope
    MainRepository provideMainRepository(CityService service) {
        return new MainRepository(service);
    }

    @Provides
    @ActivityScope
    MainPresenter provideMainPresenter(MainRepository repository) {
        return new MainPresenterImpl(view, repository);
    }


}

File: CleanCodeKotlin/app/src/main/java/io/chuumong/cleancodekotlin/di/main/MainModule.kt
package io.chuumong.cleancodekotlin.di.main

import dagger.Module
import dagger.Provides
import io.chuumong.cleancodekotlin.ui.activity.MainActivity
import io.chuumong.cleancodekotlin.data.api.CityService
import io.chuumong.cleancodekotlin.data.repository.CityRepository
import io.chuumong.cleancodekotlin.di.ActivityScope
import io.chuumong.cleancodekotlin.presenter.MainPresenter
import io.chuumong.cleancodekotlin.presenter.MainPresenterImpl

/**
 * Created by JongHunLee on 2017-02-23.
 */
@Module
class MainModule(val activity: MainActivity) {


    @Provides
    @ActivityScope
    fun provideCityRepository(service: CityService) = CityRepository(service)

    @Provides
    @ActivityScope
    fun provideMainPresenter(repository: CityRepository): MainPresenter = MainPresenterImpl(activity, repository)
}

